File: app/src/main/java/com/doapps/habits/fragments/RegisterFragment.java
package com.doapps.habits.fragments;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import com.doapps.habits.R;
import com.doapps.habits.helper.EmailTextWatcher;
import com.doapps.habits.helper.PasswordTextWatcher;
import com.doapps.habits.listeners.NameChangeListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;

/**
 * A simple {@link Fragment} subclass.
 */
public class RegisterFragment extends Fragment {

  private static void handleRegisterError(Task task, Activity activity) {
    task.addOnFailureListener(activity, fail -> {
          // Sign in failed.
          Log.e("task failed", fail.getMessage());
          Toast.makeText(
              activity.getApplicationContext(),
              fail.getMessage(),
              Toast.LENGTH_SHORT).show();
        }
    );
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    View result = inflater.inflate(R.layout.fragment_register, container, false);
    EditText inputFullName = result.findViewById(R.id.full_name);
    EditText inputEmail = result.findViewById(R.id.email);
    EditText inputPassword = result.findViewById(R.id.password);
    inputPassword.addTextChangedListener(new PasswordTextWatcher(inputPassword));
    Button btnRegister = result.findViewById(R.id.btnCreate);
    Button btnLinkToLogin = result.findViewById(R.id.btnLinkToLoginScreen);

    btnLinkToLogin.setOnClickListener(view -> toLoginActivity());
    inputEmail.setText(
        getActivity().getSharedPreferences("pref", Context.MODE_PRIVATE)
            .getString("last email", ""));
    inputEmail.addTextChangedListener(new EmailTextWatcher(inputEmail));

    FirebaseAuth mAuth = FirebaseAuth.getInstance();
    // Register Button Click event
    btnRegister.setOnClickListener(view -> {
          String name = inputFullName.getText().toString().trim();
          String email = inputEmail.getText().toString().trim();
          String password = inputPassword.getText().toString().trim();

          if (!email.isEmpty() && !password.isEmpty() && !name.isEmpty()) {
            mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(getActivity(), task -> {
                  if (task.isSuccessful()) {
                    setUserName(name);
                    toLoginActivity();
                  } else {
                    handleRegisterError(task, getActivity());
                  }
                });
          } else {
            Toast.makeText(getActivity().getApplicationContext(),
                "Please enter your details!", Toast.LENGTH_LONG)
                .show();
          }
        }

    );

    return result;
  }

  private void toLoginActivity() {
    getActivity().getSupportFragmentManager().beginTransaction()
        .setCustomAnimations(R.anim.fade_in, R.anim.fade_out)
        .replace(R.id.frame_layout, new LoginFragment()).commit();
  }

  private void setUserName(String name) {
    FirebaseUser user = FirebaseAuth.getInstance().getCurrentUser();
    if (user != null) {
      UserProfileChangeRequest.Builder changeRequest =
          new UserProfileChangeRequest.Builder();
      changeRequest.setDisplayName(name);
      user.updateProfile(changeRequest.build()).addOnCompleteListener(task -> {
        if (NameChangeListener.Companion.getListener().countObservers() == 0) {
          if (getActivity() != null) {
            NavigationView nav = getActivity().findViewById(R.id.navigationView);
            if (nav != null) {
              ((TextView) nav.getHeaderView(0).findViewById(R.id.name_info)).setText(name);
            } else {
              Log.i("updateProfile", "nav is null");
            }
          } else {
            Log.i("updateProfile", "activity is null");
          }
        } else {
          NameChangeListener.Companion.getListener().setChanged();
          Log.i("NameChangeListener", "notifyObservers");
        }
      });
    }
  }

}

File: app/src/main/java/com/doapps/habits/fragments/RegisterFragment.kt
package com.doapps.habits.fragments

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import com.doapps.habits.R
import com.doapps.habits.helper.EmailTextWatcher
import com.doapps.habits.helper.PasswordTextWatcher
import com.doapps.habits.listeners.NameChangeListener
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserProfileChangeRequest

class RegisterFragment : Fragment() {

  private fun handleRegisterError(task: Task<*>, activity: Activity?) {
    task.addOnFailureListener(activity!!
    ) { fail ->
      // Sign in failed.
      Log.e("task failed", fail.message)
      Toast.makeText(
          activity.applicationContext,
          fail.message,
          Toast.LENGTH_SHORT).show()
    }
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                            savedInstanceState: Bundle?): View? {
    // Inflate the layout for this fragment
    val result = inflater.inflate(R.layout.fragment_register, container, false)
    val inputFullName = result.findViewById<EditText>(R.id.full_name)
    val inputEmail = result.findViewById<EditText>(R.id.email)
    val inputPassword = result.findViewById<EditText>(R.id.password)
    inputPassword.addTextChangedListener(PasswordTextWatcher(inputPassword))
    val btnRegister = result.findViewById<Button>(R.id.btnCreate)
    val btnLinkToLogin = result.findViewById<Button>(R.id.btnLinkToLoginScreen)

    btnLinkToLogin.setOnClickListener { toLoginActivity() }
    inputEmail.setText(
        activity!!.getSharedPreferences("pref", Context.MODE_PRIVATE)
            .getString("last email", ""))
    inputEmail.addTextChangedListener(EmailTextWatcher(inputEmail))

    val mAuth = FirebaseAuth.getInstance()
    // Register Button Click event
    btnRegister.setOnClickListener {
      val name = inputFullName.text.toString().trim { it <= ' ' }
      val email = inputEmail.text.toString().trim { it <= ' ' }
      val password = inputPassword.text.toString().trim { it <= ' ' }

      if (!email.isEmpty() && !password.isEmpty() && !name.isEmpty()) {
        mAuth.createUserWithEmailAndPassword(email, password)
            .addOnCompleteListener(activity!!) { task ->
              if (task.isSuccessful) {
                setUserName(name)
                toLoginActivity()
              } else {
                handleRegisterError(task, activity)
              }
            }
      } else {
        Toast.makeText(activity!!.applicationContext,
            "Please enter your details!", Toast.LENGTH_LONG)
            .show()
      }
    }

    return result
  }

  private fun toLoginActivity() {
    activity!!.supportFragmentManager.beginTransaction()
        .setCustomAnimations(R.anim.fade_in, R.anim.fade_out)
        .replace(R.id.frame_layout, LoginFragment()).commit()
  }

  private fun setUserName(name: String) {
    val user = FirebaseAuth.getInstance().currentUser
    if (user != null) {
      val changeRequest = UserProfileChangeRequest.Builder()
      changeRequest.setDisplayName(name)
      user.updateProfile(changeRequest.build()).addOnCompleteListener {
        if (NameChangeListener.listener.countObservers() == 0) {
          if (activity != null) {
            val nav = activity!!.findViewById<NavigationView>(R.id.navigationView)
            if (nav != null) {
              (nav.getHeaderView(0).findViewById<View>(R.id.name_info) as TextView).text = name
            } else {
              Log.i("updateProfile", "nav is null")
            }
          } else {
            Log.i("updateProfile", "activity is null")
          }
        } else {
          NameChangeListener.listener.setChanged()
          Log.i("NameChangeListener", "notifyObservers")
        }
      }
    }
  }

}
